mlrun/run.py

```diff
@@ -1218,3 +1226,609 @@ def wait_for_runs_completion(runs: list, sleep=3, timeout=0, silent=False):
+        Log the given outputs as artifacts with the stored context.
+
+        :param outputs:              List of outputs to log.
+        :param logging_instructions: List of logging instructions to use - tuples of artifact types and keys or logging
+                                     kwargs.
+        """
+        for obj, log_tuple in zip(outputs, logging_instructions):
+            # Check if needed to log (not None):
+            if log_tuple is None:
+                continue
+            # Log:
+            self._log_output(obj=obj, artifact_type=log_tuple[0], key=log_tuple[1])
+
+    def set_labels(self, labels: Dict[str, str]):
+        """
+        Set the given labels with the stored context.
+
+        :param labels: The labels to set.
+        """
+        for key, value in labels.items():
+            self._context.set_label(key=key, value=value)
+
+    def _parse_input(self, data_item: DataItem, expected_type: Type) -> Any:
+        """
+        Parse the given data frame to the expected type. By default, it will be parsed to an object (will be treated as
+        a pickle).
+
+        :param data_item:     The data item to parse.
+        :param expected_type: THe expected type to parse to.
+
+        :returns: The parsed data item.
+        """
+        return self.PARSER_MAP.get(str(expected_type), self.PARSER_MAP[str(object)])(
+            data_item=data_item
+        )
+
+    def _log_output(
+        self, obj, artifact_type: Union[ArtifactType, str], key: Union[str, dict]
+    ):
+        """
+        Log the given object to MLRun as the given artifact type with the provided key. The key can be part of a
+        logging keyword arguments to pas to the relevant context logging function.
+
+        :param obj:           The object to log.
+        :param artifact_type: The artifact type to log the object as.
+        :param key:           The key (name) of the artifact or a logging kwargs to use when logging the artifact.
+
+        :raises MLRunInvalidArgumentError: If a key was not provided.
+        """
+        # Get the artifact type:
+        artifact_type = ArtifactType(artifact_type)
+
+        # Check if the key is in fact logging kwargs:
+        logging_kwargs = {}
+        if isinstance(key, dict):
+            logging_kwargs = key
+            # Get the key value from the kwargs:
+            key = logging_kwargs.pop("key" if "key" in logging_kwargs else "item", None)
+            if key is None:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "When passing logging keyword arguments, either 'key' or 'item' (according to the context "
+                    "method) must be added to the dictionary."
+                )
+
+        # Use the logging map to log the object:
+        self.LOGGER_MAP[artifact_type.value](
+            context=self._context,
+            obj=obj,
+            key=key,
+            logging_kwargs=logging_kwargs,
+        )
+
+
+def function_decorator(
```

Comment:
I am not sure about the decorator naming.
Looking at some information about decorators such as https://stackoverflow.com/a/47968658/19186532 and https://peps.python.org/pep-0318/.  I think we need to call the function with what it does ( Maybe `log_artifacts` not sure about the suggested naming 😅) and remove the `decorator` prefix from the function.

Reply:
It suppose to be an MLRun Swiss-army-knife, enabling multiple functionalities regarding MLRun's context, to reduce the friction in beginning to use MLRun.

So, `log_artifacts` can't be its name as it is only one of the options, it also parse `DataItem`s and set labels. In the future it can wrap ML/DL frameworks to enable auto-logging and auto-evaluations.

On one hand, you are write, `function_decorator` is kinda weird. But on the other hand, when using it as `mlrun.function_decorator` it makes sense as it is decorating an MLRun function. We can call it `function_wrapper` if you prefer, as "wrapping" is not something used in python's literature.

Reply:
Decided to use `@context` as this will wrap all context functionality eventually.